BTreeMap: Support custom allocators

exrook opened the pull request:
Related:
https://github.com/rust-lang/wg-allocators/issues/7
https://github.com/TimDiekmann/alloc-wg

Blocked on:
~~#77187~~
~~#78459~~ 
~~#95036~~

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #77436) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

exrook commented:
@rustbot modify labels: +A-allocators +T-libs

bors commented:
:umbrella: The latest upstream changes (presumably #77470) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

TimDiekmann commented:
@rustbot modify labels: -S-waiting-on-review -S-blocked

TimDiekmann commented on library/alloc/src/collections/btree/map.rs:
````diff
@@ -533,7 +578,31 @@ impl<K: Ord, V> BTreeMap<K, V> {
+    /// ```
+    /// use std::collections::BTreeMap;
+    ///
+    /// let mut map = BTreeMap::new();
````
In this example, `new_in` should be used.

TimDiekmann commented on library/alloc/src/collections/btree/map.rs:
````diff
@@ -122,13 +124,14 @@ use UnderflowResult::*;
 /// ```
 #[stable(feature = "rust1", since = "1.0.0")]
-pub struct BTreeMap<K, V> {
+pub struct BTreeMap<K, V, #[unstable(feature = "allocator_api", issue = "32838")] #[unstable(feature = "allocator_api", issue = "32838")] A: AllocRef = Global> {
````
You added `#[unstable]` twice

```suggestion
pub struct BTreeMap<K, V, #[unstable(feature = "allocator_api", issue = "32838")] A: AllocRef = Global> {
```

TimDiekmann commented:
cc @Amanieu

cramertj commented:
r? @Amanieu

bors commented:
:umbrella: The latest upstream changes (presumably #77981) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

bors commented:
:umbrella: The latest upstream changes (presumably #78060) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

bors commented:
:umbrella: The latest upstream changes (presumably #77244) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

bors commented:
:umbrella: The latest upstream changes (presumably #78904) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Dylan-DPC commented:
unblocking this as both issues this was blocked on (as per the title) have been resolved

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5493868843?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5148d228b2acf16c899db9c6975ecc5204340a81/checks/5493868843/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling sha-1 v0.10.0
   Compiling md-5 v0.10.0
   Compiling rustc-rayon v0.3.2
   Compiling tempfile v3.2.0
error: failed to parse bitcode for LTO module: Explicit gep type does not match pointee type of pointer operand (Producer: 'LLVM14.0.0-rust-1.60.0-beta' Reader: 'LLVM 14.0.0-rust-1.60.0-beta')
error: could not compile `aho-corasick` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:00:55
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5495773835?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/32d65ae9fb3fe4150ba77b86a4c19289779d6731/checks/5495773835/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling sha2 v0.10.1
   Compiling md-5 v0.10.0
   Compiling sha-1 v0.10.0
   Compiling tempfile v3.2.0
error: failed to parse bitcode for LTO module: Explicit gep type does not match pointee type of pointer operand (Producer: 'LLVM14.0.0-rust-1.60.0-beta' Reader: 'LLVM 14.0.0-rust-1.60.0-beta')
error: could not compile `aho-corasick` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:01:00
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5502192823?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a1bd1e1806f29a05aa7bab8a857be587e6c7893b/checks/5502192823/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling sha2 v0.10.1
   Compiling md-5 v0.10.0
   Compiling rustc-rayon v0.3.2
   Compiling tempfile v3.2.0
error: failed to parse bitcode for LTO module: Explicit gep type does not match pointee type of pointer operand (Producer: 'LLVM14.0.0-rust-1.60.0-beta' Reader: 'LLVM 14.0.0-rust-1.60.0-beta')
error: could not compile `aho-corasick` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:01:17
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #92962) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5642538698?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/72e48ac90e5f6da1ba0bf632e3313b844d80d564/checks/5642538698/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/alloc/src/collections/btree/map.rs at line 296:
         match root_node.search_tree::<K>(&key) {
             Found(mut kv) => Some(mem::replace(kv.key_mut(), key)),
             GoDown(handle) => {
-                VacantEntry { key, handle: Some(handle), dormant_map, alloc: &*map.alloc, _marker: PhantomData }
-                    .insert(());
+                VacantEntry {
+                    key,
+                    handle: Some(handle),
+                    dormant_map,
+                    alloc: &*map.alloc,
+                    _marker: PhantomData,
+                .insert(());
                 None
             }
         }
         }
Diff in /checkout/library/alloc/src/collections/btree/map.rs at line 1103:
         let other_iter =
             mem::replace(other, Self::new_in(ManuallyDrop::into_inner(self.alloc.clone())))
                 .into_iter();
-        let root = self.root.get_or_insert_with(||Root::new(&*self.alloc));
+        let root = self.root.get_or_insert_with(|| Root::new(&*self.alloc));
         root.append_from_sorted_iters(self_iter, other_iter, &mut self.length, &*self.alloc)
 
Diff in /checkout/library/alloc/src/collections/btree/map.rs at line 1217:
     {
     {
         let (map, dormant_map) = DormantMutRef::new(self);
         match map.root {
-            None => Vacant(VacantEntry { key, handle: None, dormant_map, alloc: &*map.alloc, _marker: PhantomData }),
+            None => Vacant(VacantEntry {
+                key,
+                dormant_map,
+                dormant_map,
+                alloc: &*map.alloc,
+                _marker: PhantomData,
+            }),
             Some(ref mut root) => match root.borrow_mut().search_tree(&key) {
-                Found(handle) => {
-                    Occupied(OccupiedEntry { handle, dormant_map, alloc: &*map.alloc, _marker: PhantomData })
-                }
+                Found(handle) => Occupied(OccupiedEntry {
+                    dormant_map,
+                    dormant_map,
+                    alloc: &*map.alloc,
+                    _marker: PhantomData,
+                }),
                 GoDown(handle) => Vacant(VacantEntry {
                     key,
                     handle: Some(handle),
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/alloc/src/collections/btree/map.rs" "/checkout/library/alloc/src/collections/vec_deque/mod.rs" "/checkout/library/alloc/src/collections/btree/remove.rs" "/checkout/library/alloc/src/collections/vec_deque/pair_slices.rs" "/checkout/library/alloc/src/collections/btree/testing/mod.rs" "/checkout/library/alloc/src/collections/btree/testing/crash_test.rs" "/checkout/library/alloc/src/collections/btree/node/tests.rs" "/checkout/library/alloc/src/collections/vec_deque/iter_mut.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:09
Diff in /checkout/library/alloc/src/collections/btree/map/tests.rs at line 116:
         let iter = mem::take(self).into_iter();
         let iter = mem::take(self).into_iter();
         if !iter.is_empty() {
-            self.root.insert(Root::new(&*self.alloc)).bulk_push(iter, &mut self.length, &*self.alloc);
+            self.root.insert(Root::new(&*self.alloc)).bulk_push(
+                iter,
+                &mut self.length,
+                &*self.alloc,
         }
     }
 }
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #95702) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

JohnCSimon commented:
Triage: 
@Amanieu is this ready for review?

Amanieu commented on src/bootstrap/builder.rs:
```diff
@@ -1684,6 +1684,7 @@ impl<'a> Builder<'a> {
                 rustflags.arg(&format!("-Cllvm-args=-import-instr-limit={}", limit));
             }
         }
+        cargo.env(profile_var("LTO"), "off");
```
Leftover from testing?

Amanieu reviewed (approved):
Sorry about the late review, this looks good modulo the leftover test code. Can you rebase to resolve the conflicts?

TennyZhuang commented:
What's the status of the PR? I'd like to help with something.

Amanieu commented:
@TennyZhuang Since the author is not responding, you can take over by forking their code and opening a new PR.

TennyZhuang commented:
@exrook Sorry but I really need the feature, I'll take over the PR at the weekend. 🥰

TennyZhuang commented:
@Amanieu Hi, I've resolved the conflicts and submitted a new PR #98015, PTAL Thanks!

Dylan-DPC commented:
Closing this in favour of #98015

exrook commented:
@TennyZhuang I'm glad to hear you're also looking to use this feature! My apologies for not responding sooner, I've been busy with other things, and when I last attempted a rebase on master back in May I was still hitting box-related ICEs so I've been working on a branch based on top of #95576, though it seems now those ICEs have been fixed in master :).

@Dylan-DPC Hi, could this be re-opened? I've put a lot of effort into keeping these changes in sync with master over the past year+ and a half and I'd like to see it through to being merged. :)

@Amanieu Thanks for the review! I've rebased on master and removed the leftover test code, so this should be good to go if you have no other comments

TennyZhuang commented:
@exrook I’m glad to hear that, and really hope the PR can be merged ASAP.
@Dylan-DPC @Amanieu Can you help to reopen the PR and review it again? Sorry for the extra review work.

Amanieu commented:
GIthub doesn't let me reopen the PR after a force-push/rebase. You'll need to open a new one.

TennyZhuang commented:
@exrook I’ve found that it’s possible to reopen the PR, but it need you to revert the commits back to the one before closed, sorry.

See https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c

Feel free to reopen the PR or just create a new one.

Dylan-DPC commented:
@exrook unfortunately you can't reöpen this. You can submit a new pull request and link this one in the description and we can take it from there

exrook commented:
Thanks everyone! I've opened a new PR at #98103
